### PR TITLE
[MoE] Support padding masks in load balancing

### PR DESCRIPTION
**Pull request by acisseJZhong**
## Summary

- propagate `padding_mask` through decoder, MoE, and MTP paths
- validate padding masks against the routing-map token axis, then build one padding-filtered routing map for expert counts and auxiliary load-balancing statistics
- preserve the original routing map for dispatch, and shard the mask with router tokens under expert parallelism

This follows the separation between execution routing and padding-filtered statistics used by the [Megatron-LM router implementation](https://github.com/NVIDIA/Megatron-LM/blob/f2f0f7bfd88fcb1243df55275988d6af52daea35/megatron/core/transformer/moe/router.py#L737-L763). The no-mask path is unchanged.

## Test results

2x H100, FP32, deterministic seed 42, TP=2, EP=2.

| Test | Before loss (steps 1, 2) | After loss (steps 1, 2) | Result |
| --- | --- | --- | --- |
| DeepSeek V3, no padding, SP | 8.06509018, 6.36459827 | 8.06509018, 6.36459827 | Exact match |
| DeepSeek V3, padding, SP | 7.92675018, 5.56190205 | 7.92675018, 5.56203842 | Step 1 exact; step 2 +0.00013638 after the padding-aware update |
| DeepSeek V3, padding, no SP | N/A | 7.92675018, 5.56203842 | Exact match with the after-SP run |
| DeepSeek V4 hash router, padding, SP | 8.15526104, 5.90886402 | 8.15526104, 5.90848827 | Step 1 exact; step 2 -0.00037575 after the padding-aware expert-bias update |

## Follow-up

- assign padding tokens to experts round-robin to balance dispatch and expert compute, while continuing to exclude them from auxiliary loss, expert-bias updates, and usage metrics

### CODE REVIEW

File: torchtitan/models/deepseek_v4/sharding.py [hunk 1]
```diff
@@ -278,20 +278,13 @@ def set_deepseek_v4_layer_sharding(
...
         router_cfg = layer_cfg.moe.router
         if getattr(router_cfg, "layer_id", 0) < getattr(router_cfg, "n_hash_layers", 0):
-            # tokens / input_ids_T enter the model TP-replicated. MoE keeps
-            # activations sequence-parallel only on the SP+EP path, so hash
-            # ids must shard T the same way. Otherwise x is all-gathered or
-            # unreplicated on TP and the full token-id tensor matches.
             input_ids_src_placement = token_id_placement()
-            input_ids_dst_placement = (
-                token_id_sequence_parallel_placement()
-                if enable_sp and enable_ep
```
Comment (acisseJZhong) [inline torchtitan/models/deepseek_v4/sharding.py:274]: this is an intentional bug fix introduced in https://github.com/pytorch/torchtitan/pull/4490/changes#diff-4470b7d1c2056fee681bb0713ed8b980cec39a894a9d83e712e949ea2b2eb09cL293

as we always enable SP for MoE 
@YeonwooSung please correct me if I am misunderstanding anything

Comment (YeonwooSung) [reply torchtitan/models/deepseek_v4/sharding.py:274]: You have it right for the configuration that is actually used, but the `enable_sp and` half is load-bearing for one combination, so `token_id_placement(enable_sp=enable_ep)` is not equivalent to what is there today.

SP is on by default but not enforced. `enable_sp` comes from `parallelism.enable_sequence_parallel` (a user-settable field, default `True`) and `enable_ep` from `expert_parallel_degree > 1`; nothing ties them together, so `--parallelism.enable_sequence_parallel=False` with `expert_parallel_degree > 1` is reachable.

That combination is exactly the one the conditional was protecting, because of `_moe_sharding_config` in `torchtitan/models/common/moe_sharding.py`:

```python
sp_layout = (
    dense_sequence_parallel_placement() if enable_sp
    else dense_activation_placement(tp=spmd.I, cp=spmd.S(0))
)
desired_input_layout = (
    sp_layout if enable_ep else dense_activation_placement(tp=spmd.R, cp=spmd.S(0))
)
```

With EP on and SP off, `x_TD`'s TP axis is `I`, not `S(0)` -- each rank still holds the full T. So the rule is not "EP decides whether x is sequence-parallel", it is "EP *and* SP together decide it". Under this PR the hash ids become `S(0)` on TP while `x_TD` stays full-T, which puts `tid2eid[input_ids_T]` back to indexing a `T/tp`-sized id tensor against a full-T activation -- the same class of mismatch #4490 fixed (CI showed `scores[256, 4]` vs `ids[512, 3]`, just from the other direction). Today's `else token_id_placement()` keeps both sides full T, so the shapes agree even though the configuration is otherwise a bit odd.

Two ways to get the simplification you want:

1. Keep the two-condition form (`enable_sp and enable_ep`). Merging `token_id_sequence_parallel_placement` into `token_id_placement(enable_sp=...)` is a nice cleanup on its own; it just wants `enable_sp=enable_sp and enable_ep` at this call site.
2. If EP without SP should not be a supported configuration, make that explicit -- reject it (or force SP) in `ParallelismConfig.__post_init__` -- and then `enable_sp=enable_ep` really is equivalent and the invariant is checked rather than assumed.

One thing to flag either way: this PR deletes `tests/unit_tests/cpu/test_deepseek_v4_hash_sharding.py` entirely, and `test_ep_without_sp_keeps_hash_input_ids_replicated_on_tp` in it is the test that covers this case. The `token_id_sequence_parallel_placement` assertions in `test_parallel_dims.py` go too, and I did not find replacement coverage for the hash-routing placement in the new tests. If the helper rename is what made the file awkward to keep, updating the imports would keep the coverage; as it stands the change to the condition and the removal of the test that would catch it land together.

Happy to be wrong about the EP-without-SP path if there is a constraint I have missed -- I could not run anything locally here, so this is from reading `moe_sharding.py` and the config, not from a repro.

File: torchtitan/models/common/decoder.py [hunk 2]
```diff
@@ -249,6 +249,8 @@ def forward(
         tokens: torch.Tensor,
         positions: torch.Tensor | None = None,
         attention_masks: AttentionMasksType | None = None,
+        *,
+        padding_mask: torch.Tensor | None = None,
```
Comment (tianyu-l) [inline torchtitan/models/common/decoder.py:239]: hoping to make non-tokens, all optional fields to kwargs, but could be a followup

File: torchtitan/models/muse_glimmer/model.py [hunk 3]
```diff
@@ -535,7 +539,12 @@ def forward(
             h_TD = tokens
 
         for layer in self.layers.values():
-            h_TD = layer(h_TD, attention_masks, positions)
+            h_TD = layer(
+                h_TD,
+                attention_masks,
+                positions,
+                padding_mask=padding_mask,
```
Comment (tianyu-l) [inline torchtitan/models/muse_glimmer/model.py]: I thought muse glimmer is dense only

File: torchtitan/models/common/moe_sharding.py [hunk 4]
```diff
@@ -270,7 +271,12 @@ def _routed_experts_sharding_configs(
     )
 
 
-def _moe_sharding_config(*, enable_ep: bool, enable_sp: bool) -> ShardingConfig:
+def _moe_sharding_config(
+    *,
+    enable_ep: bool,
+    enable_sp: bool,
+    padding_mask_sequence_sharded: bool = False,
```
Comment (tianyu-l) [inline torchtitan/models/common/moe_sharding.py]: confused by why we need this -- I had thought
- padding mask sharding should always follow token sharding
- enable_sp and enable_ep are enough to determine padding's sharding

File: torchtitan/models/deepseek_v3/mtp.py [hunk 5]
```diff
@@ -148,20 +148,27 @@ def forward(
         self,
         mtp_input_embed: torch.Tensor,
         prev_embed: torch.Tensor,
-        mtp_input_valid_mask: torch.Tensor,
+        mtp_input_valid_mask_T: torch.Tensor,
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/mtp.py]: other args don't have shape suffix

File: torchtitan/models/deepseek_v4/mtp.py [hunk 6]
```diff
@@ -68,6 +71,11 @@ def forward(  # pyrefly: ignore[bad-param-name-override]
             mtp_input_ids_T,
             attention_masks,
             positions,
+            padding_mask=(
+                ~mtp_input_valid_mask_T
+                if padding_mask_T is None
+                else ~mtp_input_valid_mask_T | padding_mask_T
+            ),
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v4/mtp.py:78]: code structure looks a bit different between v3 mtp and v4 mtp -- here it looks more succinct

File: torchtitan/models/common/moe.py [hunk 7]
```diff
@@ -364,21 +400,29 @@ def forward(
...
+                routing_map_TE
+                if padding_mask_T is None
+                else routing_map_TE & ~padding_mask_T.unsqueeze(-1)
             )
-
+            # TODO: Activation Checkpointing has the side effect of double counting tokens_per_expert_E --
+            #       first in the forward pass, and then in the backward pass. However, this has no
+            #       effect on the expert bias update thanks to the torch.sign() operator.
+            with torch.no_grad():
+                self.tokens_per_expert_E.add_(masked_routing_map_TE.sum(dim=0))
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:340]: I feel this would be helpful for inference as well -- can we move it out of if `self.training`?

File: torchtitan/models/deepseek_v4/moe.py
Comment (tianyu-l) [inline torchtitan/models/deepseek_v4/moe.py:1]: need to rebase onto https://github.com/pytorch/torchtitan/pull/4631 and fix deepseekv3 as well

File: torchtitan/models/common/moe.py [hunk 8]
```diff
@@ -191,6 +191,23 @@ def parallelize(self, parallel_dims) -> None:
         )
 
 
+def _validate_padding_mask(
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: nit: let's put this inline to make the boundary clear

File: torchtitan/models/common/moe.py [hunk 9]
```diff
@@ -504,6 +550,8 @@ def forward(
             # scores.  F.normalize's eps clamp only guards an all-zero score
             # row: the scores are non-negative, so the norm is a plain sum.
             probs_TE = F.normalize(scores_TE, p=1, dim=-1)
+            if padding_mask_T is not None:
+                probs_TE = probs_TE * ~padding_mask_T.unsqueeze(-1)
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:476]: after this, I believe we are making this line legit https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L912

see discussions in https://github.com/pytorch/torchtitan/pull/3864#discussion_r3964135109

cc @sdmyzlp

File: torchtitan/models/deepseek_v3/mtp.py
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/mtp.py:277]: I think we need to roll the `padding_mask` as well?

Comment (tianyu-l) [reply torchtitan/models/deepseek_v3/mtp.py:277]: In fact, I think for MTP https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L912 is invalid for MTP modules' `AuxLoss`, as each MTP module should have its own `global_valid_tokens`. Not sure what's the most clean way to fix.

cc @sdmyzlp @Matrix-Z97

**Review by tianyu-l — COMMENTED** [review]
very nice, identified some loose ends.

File: torchtitan/models/common/moe.py (hunk 7 above)
Comment (acisseJZhong) [reply torchtitan/models/common/moe.py:340]: https://github.com/pytorch/torchtitan/pull/4148

we don't want to update during inference/eval as it could potentially leak into the next step's expert_bias_E update?

Comment (tianyu-l) [reply torchtitan/models/common/moe.py:340]: expert_bias_E is never updated in inference

File: torchtitan/models/common/moe_sharding.py (hunk 4 above)
Comment (acisseJZhong) [reply torchtitan/models/common/moe_sharding.py]: I think this is for MTP only, added comment 
```
    # Regular decoder layers pass the TP-replicated model input mask directly
    # to MoE. MTP redistributes the mask to sequence-sharded before calling its
    # nested MoE. This flag describes the mask layout at the MoE boundary so an
    # already-sharded MTP mask is not sharded again.
```

File: torchtitan/models/deepseek_v3/mtp.py (hunk 5 above)
Comment (acisseJZhong) [reply torchtitan/models/deepseek_v3/mtp.py]: yeah I thought it would be more easier to understand if we add shape for mask, especially when doing mask manipulation. 

let me remove, since we don't have shape annotations for things like positions and prev_embed anyways.

File: torchtitan/models/deepseek_v3/mtp.py
Comment (acisseJZhong) [reply torchtitan/models/deepseek_v3/mtp.py:277]: yes I missed that. Now passed in padding mask and it will update the `mtp_input_valid_mask`.

File: torchtitan/models/deepseek_v4/mtp.py (hunk 6 above)
Comment (acisseJZhong) [reply torchtitan/models/deepseek_v4/mtp.py:78]: V4 delegates the full block to super().forward(), while V3 implements the MoE branch inline. The mask semantics are identical.

File: torchtitan/models/common/moe_sharding.py (hunk 4 above)
Comment (tianyu-l) [reply torchtitan/models/common/moe_sharding.py]: I didn't get this. In my mind MTP treatment wouldn't be very different from regular decoder. Could you give an example where they have to be different.

File: torchtitan/models/common/moe.py (hunk 7 above)
Comment (acisseJZhong) [reply torchtitan/models/common/moe.py:340]: yes but I thought #4148 introduces this bug fix to prevent accumulating this value during eval, which will pollute the next optimizer.update.

File: torchtitan/models/common/moe_sharding.py (hunk 4 above)
Comment (acisseJZhong) [reply torchtitan/models/common/moe_sharding.py]: > padding mask sharding should always follow token sharding

I think you are right. Previously we have tokens S(0) but padding mask R for MoE blocks, and for MTP will redistribute padding mask to S(0). So for MTP layers, I need a way to specify that the padding mask is already sharded to avoid double sharding. 

Now I changed the padding mask to be S(0) at MoE block level, same as tokens, by introducing `set_moe_block_padding_mask_sharding`.

File: torchtitan/models/common/moe.py (hunk 7 above)
Comment (tianyu-l) [reply torchtitan/models/common/moe.py:340]: oh ... right, that's when it happens on the same instance. OK I agree with keeping it not accumulated for inference for now.

**Review by tianyu-l — APPROVED** [review]
regarding https://github.com/pytorch/torchtitan/pull/4594#discussion_r3997915031, at least add a NOTE / TODO around https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L912

File: torchtitan/models/deepseek_v3/mtp.py
Comment (acisseJZhong) [reply torchtitan/models/deepseek_v3/mtp.py:277]: assigned the todo to @sdmyzlp

Comment (tianyu-l) [reply torchtitan/models/deepseek_v3/mtp.py:277]: @acisseJZhong where did you assign it? I was hoping on the trainer `AuxLoss.set_step_denominator` call.

Could you also make a GH issue?

Comment (acisseJZhong) [reply torchtitan/models/deepseek_v3/mtp.py:277]: added it here. also created https://github.com/pytorch/torchtitan/issues/4682

Image

**Comment by acisseJZhong** [comment]
landing this PR, the failing test https://github.com/pytorch/torchtitan/actions/runs/34900521241/job/104165936715?pr=4594 seems to be pre-existing. See https://github.com/pytorch/torchtitan/pull/4535#issuecomment-5671674241

File: torchtitan/models/deepseek_v3/mtp.py
Comment (sdmyzlp) [reply torchtitan/models/deepseek_v3/mtp.py:277]: > In fact, I think for MTP https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L912 is invalid for MTP modules' `AuxLoss`, as each MTP module should have its own `global_valid_tokens`. Not sure what's the most clean way to fix.
> 
> cc @sdmyzlp @Matrix-Z97

Got it, will take a look at this.

**Comment by rakkit** [comment]
hey, i have a question @acisseJZhong.

should we consider extend this "padding mask" to, in general, "no-loss-tokens"? 
for case like sft, the user prompts usually not count in loss, and it is not "padding token".  for me it is a bit weird to update expert's load balance for bias or loss.   from our experiment, exclude these tokens for load-balance helped a lot in SFT.  publicly i only see times ago a russian model discussed about this.

**Comment by acisseJZhong** [comment]
> exclude these tokens for load-balance helped a lot in SFT.

@rakkit do you know if others have this similar conclusion, any reports? or any other framework does this? 

It makes sense to exclude padding tokens from load balance as they are meaningless, but prompt tokens are meaningful?
